x/gamm/twap/logic_test.go

```diff
@@ -0,0 +1,206 @@
+	tests := map[string]struct {
+		record          types.TwapRecord
+		interpolateTime time.Time
+		expRecord       types.TwapRecord
+	}{
+		"0accum": {
+			record:          newRecord(time.Unix(1, 0), sdk.NewDec(10), zeroDec, zeroDec),
+			interpolateTime: time.Unix(2, 0),
+			expRecord:       newExpRecord(OneSec.MulInt64(10), OneSec.QuoInt64(10)),
+		},
+		"small starting accumulators": {
+			record:          newRecord(time.Unix(1, 0), sdk.NewDec(10), oneDec, twoDec),
+			interpolateTime: time.Unix(2, 0),
+			expRecord:       newExpRecord(oneDec.Add(OneSec.MulInt64(10)), twoDec.Add(OneSec.QuoInt64(10))),
+		},
+		"larger time interval": {
+			record:          newRecord(time.Unix(11, 0), sdk.NewDec(10), oneDec, twoDec),
+			interpolateTime: time.Unix(55, 0),
+			expRecord:       newExpRecord(oneDec.Add(OneSec.MulInt64(44*10)), twoDec.Add(OneSec.MulInt64(44).QuoInt64(10))),
+		},
+		"same time": {
+			record:          newRecord(time.Unix(1, 0), sdk.NewDec(10), oneDec, twoDec),
+			interpolateTime: time.Unix(1, 0),
+			expRecord:       newExpRecord(oneDec, twoDec),
+		},
+		// TODO: Overflow tests
+	}
+
+	for name, test := range tests {
+		t.Run(name, func(t *testing.T) {
+			// correct expected record based off copy/paste values
+			test.expRecord.Time = test.interpolateTime
+			test.expRecord.P0LastSpotPrice = test.record.P0LastSpotPrice
+			test.expRecord.P1LastSpotPrice = test.record.P1LastSpotPrice
+
+			gotRecord := twap.InterpolateRecord(test.record, test.interpolateTime)
+			require.Equal(t, test.expRecord, gotRecord)
+		})
+	}
+}
+
+func (s *TestSuite) TestUpdateTwap() {
+	poolId := s.PrepareUni2PoolWithAssets(defaultUniV2Coins[0], defaultUniV2Coins[1])
+	newSp := sdk.OneDec()
+
+	tests := map[string]struct {
+		record     types.TwapRecord
+		updateTime time.Time
+		expRecord  types.TwapRecord
+	}{
+		"0 accum start": {
+			record:     newRecord(time.Unix(1, 0), sdk.NewDec(10), zeroDec, zeroDec),
+			updateTime: time.Unix(2, 0),
+			expRecord:  newExpRecord(OneSec.MulInt64(10), OneSec.QuoInt64(10)),
+		},
+	}
+	for name, test := range tests {
+		s.Run(name, func() {
+			// setup common, block time, pool Id, expected spot prices
+			s.Ctx = s.Ctx.WithBlockTime(test.updateTime.UTC())
+			test.record.PoolId = poolId
+			test.expRecord.PoolId = poolId
+			test.expRecord.P0LastSpotPrice = newSp
+			test.expRecord.P1LastSpotPrice = newSp
+			test.expRecord.Height = s.Ctx.BlockHeight()
+			test.expRecord.Time = s.Ctx.BlockTime()
+
+			newRecord := s.twapkeeper.UpdateRecord(s.Ctx, test.record)
+			s.Require().Equal(test.expRecord, newRecord)
+		})
+	}
+}
+
+// TestComputeArithmeticTwap tests ComputeArithmeticTwap on various inputs.
+// The test vectors are structured by setting up different start and records,
+// based on time interval, and their accumulator values.
+// Then an expected TWAP is provided in each test case, to compare against computed.
+func TestComputeArithmeticTwap(t *testing.T) {
+	denom0, denom1 := "token/B", "token/A"
```

Comment:
Can we make these global and reuse them across all twap tests, please? I see these being hardcoded in other tests